### ISSUE TITLE
Freeze on invalid string in validationSchema

### ISSUE THREAD

**Issue by lokomass** [jquense/yup#1303]
**Describe the bug**
When I remove the arobase from my email string, the app freeze.

**To Reproduce**

https://codesandbox.io/s/black-paper-34v4g?file=/src/App.js

**Expected behavior**

Error message should be shown

**Platform (please complete the following information):**
 - Browser chrome

**Comment by jquense** [comment]
please simplify this to JUST yup, i don't provide support for Formik.

**Comment by lokomass** [comment]
Thanks for your reply, but I don't  how can I separate both :(

**Comment by jquense** [comment]
follow the original template, you need to show the issue with _just_ the schema, if it only occurs in the context of a Formik form, it's likely a formik issue

**Comment by lokomass** [comment]
Here is the problem without formik : https://codesandbox.io/s/yup-without-formik-forked-p4f1d?file=/src/index.js
Thanks for your help

**Comment by jquense** [comment]
thanks for the code, I can't reproduce the freeze though, what should i be trying?

**Comment by lokomass** [comment]
You can just remove the "@" char.

**Comment by jquense** [comment]
wow this is weird. The problem is actually your regex. It seems to take a super long time to run `/^(?=.{6,100}$)\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/.test("omptable_rqsditoghjuretlapalyqsduyume.com")`.

**Comment by lokomass** [comment]
Habitually it works...  the problem appears only when the mail value NOT match regex. If correct is true, there is no problem.
I use it on other apps without yup, and I do not face this problem :(

**Comment by jquense** [comment]
> I use it on other apps without yup

I'm not sure how that would work, I can reproduce this entirely without yup, by doing:

```js
/^(?=.{6,100}$)\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/.test("omptable_rqsditoghjuretlapalyqsduyume.com")
```

How are you using it elsewhere that avoid this?

**Comment by is-jonreeves** [comment]
Also just ran into this issue while using the `.email()` validator. With DevTools open it freezes for a lot longer.

The freeze time seems to increase based on the length of the string. Using the following test strings, the freeze gets progressively longer and longer:
```
"abcdefghijklmnopqrstuvwxyzabcd
"abcdefghijklmnopqrstuvwxyzabcde
"abcdefghijklmnopqrstuvwxyzabcdef
"abcdefghijklmnopqrstuvwxyzabcdefg
```

I took a look at the [rEmail](https://github.com/jquense/yup/blob/acbb8b4f3c24ceaf65eab09abaf8e086a9f11a73/src/string.ts#L9) regex and ran it isolated in the console, with `rEmail.test('"abcdefghijklmnopqrstuvwxyzabcdefg')` the CPU hits 100% and locks up for around 10seconds. It would seem that the regex as it is susceptible to ReDoS attacks.

I was hoping to pop in a simple gating test before `.email()` to fail-fast, but that didn't seem to work for me unfortunately. I've instead resorted to creating a custom email rule with a different regex pattern that seems to perform better:

```
const rEmailCustom = /^[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i;

Yup.addMethod<StringSchema>(Yup.string, 'emailCustom', function (message?: string) {
  return this.matches(rEmailCustom, {
    name: 'emailCustom',
    message: message || 'Invalid Email Address',
    excludeEmptyString: true,
  });
});

...

email: Yup
  .string()
  .trim()
  .required()
  .emailCustom(),
```

Hope it helps anyone else running into the same issue.

**Comment by aviadhahami** [comment]
Hi, budging in as well.

#### The problem
the latter matcher of the email prefix in the regex, namely this bit:
```js
((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))
```
is too broad and introduces an infinite loop given `"`(==`\x22`) as the first character.
The regex engine is trying to match repeatedly and can't decide if it "matched" or not and due to greedy quantifiers it "goes back" and attempts re-matching. 

In order to reproduce, you can attempt and email check with `"` followed by `a`s. (see code example at bottom)

#### Debugging
This is the problematic bit in the regex101 debugger https://regex101.com/r/VYYo7o/1/debugger. Take a look at how many steps it takes for the engine to attempt matching.
> you can close the debugger on the top left and modify the input

Time consumption wise, a `"` followed by 25 `a`s will take almost 1 sec to calculate (a falsy answer), 26 `a`s will take almost 2 seconds and the time consumption will continue growing exponentially wrt the input length.

#### Security
As @is-jonreeves mentioned, using this module to test user input in an app renders the app vulnerable to ReDos attacks.

#### PoC
```js
import * as yup from 'yup';

let schema = yup.object().shape({
  email: yup.string().email(),
});

schema
  .isValid({
    email: process.argv[2],
  })
  .then(console.log)
  .catch(console.error)
```
Run with 
`$ node poc.js \"aaaaaaaaaaaaaaaaaaaaaaaaa`

**Comment by aviadhahami** [comment]
Taking another look - this should be CVEd.
See this app for an example of a vulnerable app due to this:
https://github.com/decred/politeiagui/blob/master/src/containers/User/VerifyEmail/hooks.js#L44

**Comment by is-jonreeves** [comment]
@aviadhahami thanks for digging into this and breaking down the root of the issue.

Although this seems like maybe a fringe case, I actually ran into this issue doing what I'd consider a pretty common action. I copied and pasted an email out of outlook which by default uses the format `"Firstname Lastname" <email>`. As I started deleting characters (the second `"`) everything froze. I noted this didn't appear to be as bad when the name and emails were shorter. I just distilled it down to the use of `"` being the root cause. I figured the context of my discovery was probably worth knowing as it didn't seem like that unusual of an action.
